openstack-api-charm-creation-guide.md

```diff
@@ -0,0 +1,240 @@
+```
+
+The [openstack-api layer](https://github.com/openstack-charmers/charm-layer-openstack-api) defines a series of config options and interfaces which are mostly common accross Openstack API services e.g. including the openstack-api-layer will pull in the Keystone and Mysql interfaces (among others) as well as the charm layers the new Congress charm can leverage. To instruct "charm build" to pull in the openstack-api layer edit layer.yaml:
+
+```yaml
+includes: ['layer:openstack-api']
+```
+
+# Add Congress configuration
+
+## Define Congress attributes
+
+To define the attributes of Congress, inherit the OpenStackCharm class and set Congress specific variables. This is done on the charm/lib/charm/openstack/congress.py file.
+
+```
+from charm.openstack.ip import PUBLIC, INTERNAL, ADMIN
+from charm.openstack.charm import OpenStackCharmFactory, OpenStackCharm
+from charm.openstack.adapters import OpenStackRelationAdapters
+
+class CongressCharm(OpenStackCharm):
+    # Packages the service needs installed
+    packages = ['congress-server', 'congress-common', 'python-antlr3', 'python-pymysql']
+
+    # Init services the charm manages
+    services = ['congress-server']
+
+    # Standard interface adapters class to use.
+    adapters_class = OpenStackRelationAdapters
+
+    # Ports that need exposing.
+    default_service = 'congress-api'
+    api_ports = {
+        'congress-api': {
+            PUBLIC: 1789,
+            ADMIN: 1789,
+            INTERNAL: 1789,
+        }
+    }
+
+    # Database sync command used to initalise the schema.
+    sync_cmd = ['congress-db-manage', '--config-file', '/etc/congress/congress.conf', 'upgrade', 'head']
+
+    # The restart map defines which services should be restarted when a given file changes
+    restart_map = {
+        '/etc/congress/congress.conf': ['congress-server'],
+        '/etc/congress/api-paste.ini': ['congress-server'],
+        '/etc/congress/policy.json': ['congress-server'],
+    }
+```
+
+Finally create a charm factory, this is used to allow a different Charm class to be used for different Openstack releases. In this case the same charm class will be used for all Liberty and above releases.
+
+```
+class CongressCharmFactory(OpenStackCharmFactory):
+
+    releases = {
+        'liberty': CongressCharm
+    }
+
+    first_release = 'liberty'
+```
+
+## Add Congress code to react to events
+
+### Install Congress Packages
+
+The reactive framework is going to emit events that the Congress charm can react to. the charm needs to define how its going to react to these events and also raise new events as needed.
+
+The first action a charm needs to do is to install the Congress code. This is by done running the install method from CongressCharm created earlier. Edit charm/reactive/handlers.py ...
+
+```
+import charms.reactive as reactive
+import charm.openstack.congress as congress
+import charmhelpers.contrib.openstack.utils as ch_utils
+
+charm = None
+
+def get_charm():
```

Comment:
I think `get_charm()` should be part of the library rather than in the handlers.py.  i.e. it's still about the plumbing of the charm, rather than anything to do with the handler.